.travis.yml

```diff
@@ -24,7 +24,13 @@ install:
 script:
   - docker run ${DOCKER_REPOSITORY}:${TRAVIS_COMMIT} /bin/bash -c "./scripts/build.py -t"
   - docker run ${DOCKER_REPOSITORY}:${TRAVIS_COMMIT} /bin/bash -c "./scripts/build.py -t --python-path python3"
+
+after_success:
   - mkdir -p $(dirname ${DOCKER_CACHE_FILE}) ; docker save $(docker history -q ${DOCKER_REPOSITORY}:${TRAVIS_COMMIT} | grep -v '<missing>') | gzip > ${DOCKER_CACHE_FILE};
+  - if [ "$TRAVIS_BRANCH" == "master" ]; then
+    docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD";
```

Comment:
What upsets you? From my understanding of Andrew's suggestion, it would make sense to build base images which are used at different stages:
Testing/CI
Demo in Jupiter
flask 

The most efficient way to save intermediate steps and load them to build the next image is via a public repo. 

What do you think?